org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -480,23 +480,47 @@ public synchronized void stop() {
 	 * @param document
 	 * @return null if not connection has happened, a future tracking the connection state otherwise
 	 * @throws IOException
+	 * @deprecated use {@link #connect(IDocument, IFile)}
 	 */
+	@Deprecated(forRemoval = true)
 	public @Nullable CompletableFuture<LanguageServer> connect(@NonNull IFile file, IDocument document)
 			throws IOException {
+		CompletableFuture<@NonNull LanguageServerWrapper> connect = connect(document, file);
+		if (connect != null) {
+			return connect.thenApply(theVoid -> languageServer);
+		}
+		return null;
+	}
+
+	public @Nullable CompletableFuture<@NonNull LanguageServerWrapper> connect(IDocument document, @NonNull IFile file)
+			throws IOException {
 		final URI uri = LSPEclipseUtils.toUri(file);
-		return uri == null ? null : connect(uri, document);
+		if (uri != null) {
+			return connect(uri, document);
+		}
+		return null;
 	}
 
 	/**
 	 * @param document
 	 * @return null if not connection has happened, a future tracking the connection state otherwise
 	 * @throws IOException
+	 * @deprecated use {@link #connect2}
 	 */
+	@Deprecated(forRemoval = true)
 	public @Nullable CompletableFuture<LanguageServer> connect(IDocument document) throws IOException {
+		CompletableFuture<LanguageServerWrapper> connect2 = connect2(document);
+		if (connect2 != null) {
+			return connect2.thenApply(theVoid -> languageServer);
+		}
+		return null;
+	}
+
+	public @Nullable CompletableFuture<LanguageServerWrapper> connect2(IDocument document) throws IOException {
```

Comment:
since java does not support overloading return values, I could not find anything better than this. In the next release we can deprecate this one and just change the return value of connect.

Reply:
or we can just call it `connectDocument` ?

Reply:
good idea. I have renamed it.